extensions/amp-analytics/0.1/events.js

```diff
@@ -1500,35 +1504,98 @@ export class VisibilityTracker extends EventTracker {
       // false missed searches.
       const selectionMethod =
         config['selectionMethod'] || visibilitySpec['selectionMethod'];
-      unlistenPromise = this.root
-        .getAmpElement(
-          context.parentElement || context,
-          selector,
-          selectionMethod
-        )
-        .then(element => {
-          return visibilityManagerPromise.then(
-            visibilityManager => {
-              return visibilityManager.listenElement(
-                element,
-                visibilitySpec,
-                this.getReadyPromise(waitForSpec, selector, element),
-                createReportReadyPromiseFunc,
-                this.onEvent_.bind(this, eventType, listener, element)
-              );
-            },
-            () => {}
+
+      if (multiSelectorVisibilityOn && Array.isArray(selector)) {
+        const promises = [];
+        for (let i = 0; i < selector.length; i++) {
+          promises.push(
+            this.getUnlistenPromiseForSelector_(
+              context,
+              selector[i],
+              selectionMethod,
+              visibilityManagerPromise,
+              visibilitySpec,
+              waitForSpec,
+              createReportReadyPromiseFunc,
+              eventType,
+              listener
+            )
           );
-        });
+        }
+        unlistenPromise = Promise.all(promises);
+      } else {
+        unlistenPromise = this.getUnlistenPromiseForSelector_(
+          context,
+          selector,
+          selectionMethod,
+          visibilityManagerPromise,
+          visibilitySpec,
+          waitForSpec,
+          createReportReadyPromiseFunc,
+          eventType,
+          listener
+        );
+      }
     }
 
     return function() {
       unlistenPromise.then(unlisten => {
-        unlisten();
+        if (multiSelectorVisibilityOn && Array.isArray(unlisten)) {
+          for (let i = 0; i < unlisten.length; i++) {
+            unlisten[i]();
+          }
+        } else {
+          unlisten();
+        }
       });
     };
   }
 
+  /**
+   * @param {!Element} context
+   * @param {string} selector
+   * @param {string} selectionMethod
+   * @param {Promise<!./visibility-manager.VisibilityManager>} visibilityManagerPromise
+   * @param {!JsonObject} visibilitySpec
+   * @param {string} waitForSpec
+   * @param {?function()} createReportReadyPromiseFunc
+   * @param {string} eventType
+   * @param {function(!AnalyticsEvent)} listener
+   * @return {!Promise<!UnlistenDef>}
+   */
+  getUnlistenPromiseForSelector_(
```

Comment:
haha i think i get why you want to inline this function, so many parameters that I'm starting to think if we should simply get rid of the helper function and inline everything...

One question is `getUnlistenPromiseForSelector_` currently returns one promise for each selector in the array. What does it return when you add support to `querySelectorAll` in the following PR? Still one promise, but a `Promise.all()`? 

I have an idea to not handle promise in the `selector` level, but in the `element` level, what do you think of that?

Reply:
> I have an idea to not handle promise in the selector level, but in the element level, what do you think of that?

In the next PR, I change `getUnlistenPromiseForSelector_` to `getUnlistenPromiseForElement_`, just like your idea. I'll change it to that version now.